presto/dataset.py

```diff
@@ -287,22 +361,79 @@ def __init__(
+                neg_indices, pos_indices = [], []
+                for loc_idx, (_, row) in enumerate(self.df.iterrows()):
+                    target = self.target_crop(
+                        row, self.task_type, self.croptype_list, return_hierarchical_labels
+                    )
+                    if target == 0:
+                        neg_indices.append(loc_idx)
+                    else:
+                        pos_indices.append(loc_idx)
+                if len(pos_indices) > len(neg_indices):
+                    self.indices = (
+                        pos_indices + (len(pos_indices) // len(neg_indices)) * neg_indices
+                    )
+                elif len(neg_indices) > len(pos_indices):
+                    self.indices = (
+                        neg_indices + (len(neg_indices) // len(pos_indices)) * pos_indices
+                    )
                 else:
-                    pos_indices.append(loc_idx)
-            if len(pos_indices) > len(neg_indices):
-                self.indices = pos_indices + (len(pos_indices) // len(neg_indices)) * neg_indices
-            elif len(neg_indices) > len(pos_indices):
-                self.indices = neg_indices + (len(neg_indices) // len(pos_indices)) * pos_indices
-            else:
-                self.indices = neg_indices + pos_indices
+                    self.indices = neg_indices + pos_indices
+            if self.task_type == "croptype":
+                classes_lst = self.df["balancing_class"].unique()
+
+                # optimal_class_size = self.df["balancing_class"].value_counts().max()
+                optimal_class_size = 100
+                balancing_coeff = 1.5
+
+                balanced_inds = []
+                for tclass in classes_lst:
+                    tclass_sample_ids = self.df[self.df["balancing_class"] == tclass].index
+                    tclass_loc_idx = [self.df.index.get_loc(xx) for xx in tclass_sample_ids]
+                    if len(tclass_loc_idx) < optimal_class_size:
+
+                        if balancing_coeff > 0:
+                            if (optimal_class_size / len(tclass_loc_idx)) > balancing_coeff:
+                                samples_to_add = np.random.choice(
+                                    tclass_loc_idx, size=int(len(tclass_loc_idx) / balancing_coeff)
+                                )
+                                tclass_loc_idx.extend(list(samples_to_add))
+                            else:
+                                tclass_loc_idx = tclass_loc_idx * (
+                                    optimal_class_size // len(tclass_loc_idx)
+                                )
+                        else:
+                            tclass_loc_idx = tclass_loc_idx * (
+                                optimal_class_size // len(tclass_loc_idx)
+                            )
+
+                    balanced_inds.extend(tclass_loc_idx)
+                self.indices = balanced_inds
```

Comment:
can we add a logging here that reports how many samples of each class are being repeated? So we can trace this afterwards and also spot any unexpected behavior more easily.